### ISSUE TITLE
[Question] ScreenDuplicator and FullScreenQuad? ELI5. 

### ISSUE THREAD

**Issue by lokimeyburg** [mellinoe/veldrid#94]
I'm looking through the NeoDemo and I'm curious about these lines of code (in `NeoDemo.cs`)

```
ScreenDuplicator duplicator = new ScreenDuplicator();
_scene.AddRenderable(duplicator);

_fsq = new FullScreenQuad();
 _scene.AddRenderable(_fsq);
```

What is the purpose of the ScreenDuplicator and the FullScreenQuad? How are they being used in the demo?

Commenting them out just creates a blank screen, so I'm curious about what's happening behind the scenes. 

Thanks.

**Comment by mellinoe** [comment]
Hey @lokimeyburg, this is a good question.

Like some other things in NeoDemo, this stuff is sort of a "functional test" for a couple of different features. Ultimately, it doesn't do anything very interesting or visually impressive, but it lets NeoDemo use more features and gives me a bit more confidence that I haven't broken anything.

Everything in NeoDemo is rendered to an offscreen Framebuffer. Then, that Framebuffer is "duplicated" into another Framebuffer with two Textures. This is what the screen duplicator does:

```
_outputColor_0 = clamp(texture(sampler2D(SourceTexture, SourceSampler), fsin_0), 0, 1);
_outputColor_1 = clamp(texture(sampler2D(SourceTexture, SourceSampler), fsin_0) * vec4(1.0f, 0.7f, 0.7f, 1.f), 0, 1);
```

`SourceTexture` is the texture containing the full rendered scene.

The first color it outputs is unchanged -- it just copies the rendered scene over unchanged into the first output Texture.

The second color is tinted slightly red, and its copied over to a second output Texture.

FullScreenQuad simply samples a source Texture and blits it into another Texture. In the NeoDemo "Settings" menu, you can check `Tinted output` to change which Texture it samples from. If `Tinted output` is unchecked, then it samples from the first output Texture of the duplicator -- the unchanged scene color. If `Tinted output` is checked, then the FullScreenQuad instead samples from the second duplicated Texture, the one that has been tinted slightly red.

FullScreenQuad renders to the application's main Swapchain, so its output is what actually gets drawn on screen. If you comment it out, then nothing ever gets drawn, and you get a blank screen.

If all of that seems a bit convoluted and unnecessary, you might be right 😄 . Like I alluded to above, some of the things that NeoDemo does are a bit "forced", for the sake of adding extra features and coverage.

**Comment by lokimeyburg** [comment]
Ah Ok  That makes a little more sense. The names weren't very clear. Thanks for explaining. I'll let you know if I have any more questions down the road.
